File: packages/js-dpp/lib/stateTransition/errors/PublicKeySecurityLevelNotMetError.js
const DPPError = require('../../errors/DPPError');

class PublicKeySecurityLevelNotMetError extends DPPError {
  /**
   *
   * @param {number} publicKeySecurityLevel
   * @param {number} requiredSecurityLevel
   */
  constructor(publicKeySecurityLevel, requiredSecurityLevel) {
    super(`Invalid key security level ${publicKeySecurityLevel}. This state transition requires at least ${requiredSecurityLevel}`);

    this.publicKeySecurityLevel = publicKeySecurityLevel;
    this.requiredSecurityLevel = requiredSecurityLevel;
  }

  /**
   * Get mismatched public key
   *
   * @return {number}
   */
  getPublicKeySecurityLevel() {
    return this.publicKeySecurityLevel;
  }

  /**
   * Get minimal required key security level
   *
   * @returns {number}
   */
  getKeySecurityLevelRequirement() {
    return this.requiredSecurityLevel;
  }
}

module.exports = PublicKeySecurityLevelNotMetError;

File: packages/rs-dpp/src/state_transition/errors/public_key_security_level_not_met_error.rs
use thiserror::Error;

use crate::identity::identity_public_key::SecurityLevel;
use crate::errors::ProtocolError;

#[derive(Error, Debug, Clone, PartialEq, Eq)]
#[error("Invalid key security level: {public_key_security_level}. The state transition requires at least: {required_security_level}")]
#[ferment_macro::export]
pub struct PublicKeySecurityLevelNotMetError {
    pub public_key_security_level: SecurityLevel,
    pub required_security_level: SecurityLevel,
}

impl PublicKeySecurityLevelNotMetError {
    pub fn new(
        public_key_security_level: SecurityLevel,
        required_security_level: SecurityLevel,
    ) -> Self {
        Self {
            public_key_security_level,
            required_security_level,
        }
    }

    pub fn public_key_security_level(&self) -> SecurityLevel {
        self.public_key_security_level
    }
    pub fn required_security_level(&self) -> SecurityLevel {
        self.required_security_level
    }
}

impl From<PublicKeySecurityLevelNotMetError> for ProtocolError {
    fn from(err: PublicKeySecurityLevelNotMetError) -> Self {
        Self::PublicKeySecurityLevelNotMetError(err)
    }
}
